Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -239,21 +250,257 @@ FieldValue Serializer::DecodeFieldValue(const uint8_t* bytes, size_t length) {
 
   switch (tag) {
     case google_firestore_v1beta1_Value_null_value_tag:
-      DecodeNull(&stream);
+      DecodeNull(stream);
       return FieldValue::NullValue();
     case google_firestore_v1beta1_Value_boolean_value_tag:
-      return FieldValue::BooleanValue(DecodeBool(&stream));
+      return FieldValue::BooleanValue(DecodeBool(stream));
     case google_firestore_v1beta1_Value_integer_value_tag:
-      return FieldValue::IntegerValue(DecodeInteger(&stream));
+      return FieldValue::IntegerValue(DecodeInteger(stream));
     case google_firestore_v1beta1_Value_string_value_tag:
-      return FieldValue::StringValue(DecodeString(&stream));
+      return FieldValue::StringValue(DecodeString(stream));
+    case google_firestore_v1beta1_Value_map_value_tag:
+      return FieldValue::ObjectValue(DecodeObject(stream));
 
     default:
       // TODO(rsgowman): figure out error handling
       abort();
   }
 }
 
+void EncodeSubFieldValue(pb_ostream_t* stream, const FieldValue& field_value) {
```

Comment:
I realize none of this is duplicate code but you've got a lot of things in here with very similar names. On the encoding side:

  * `EncodeFieldValueImpl`
  * `EncodeSubFieldValue`
  * `EncodeFieldEntry`
  * `EncodeFieldValue`

Perhaps `EncodeFieldEntry` is actually about maps? We need better names in here to keep these straight.

`Impl` vs not `Impl` rarely helps the reader understand what's going on. Maybe `EncodeFieldValueToStream`?

Alternatively, should we reserve `Encode` for creating nanopb struct things and come up with some other verb for writing into streams? This would also be somewhat easier to comprehend if these weren't all free functions in the same file.

Given what I can see here so far, my suggestion would be to put these things that operate on a pb_ostream into a class we can stack allocate where we need it.